tests/components/tasmota/test_common.py

```diff
@@ -0,0 +1,328 @@
+
+    data1 = json.dumps(config1)
+    data2 = json.dumps(config2)
+    assert config1[CONF_MAC] == config2[CONF_MAC]
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config1[CONF_MAC]}/config", data1)
+    await hass.async_block_till_done()
+
+    # Verify device and entity registry entries are created
+    device_entry = device_reg.async_get_device(set(), {("mac", config1[CONF_MAC])})
+    assert device_entry is not None
+    entity_entry = entity_reg.async_get(f"{domain}.test")
+    assert entity_entry is not None
+
+    # Verify state is added
+    state = hass.states.get(f"{domain}.test")
+    assert state is not None
+    assert state.name == "Test"
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config2[CONF_MAC]}/config", data2)
+    await hass.async_block_till_done()
+
+    # Verify entity registry entries are cleared
+    device_entry = device_reg.async_get_device(set(), {("mac", config2[CONF_MAC])})
+    assert device_entry is not None
+    entity_entry = entity_reg.async_get(f"{domain}.test")
+    assert entity_entry is None
+
+    # Verify state is removed
+    state = hass.states.get(f"{domain}.test")
+    assert state is None
+
+
+async def help_test_discovery_update_unchanged(
+    hass, mqtt_mock, caplog, domain, config, discovery_update
+):
+    """Test update of discovered component without changes.
+
+    This is a test helper for the MqttDiscoveryUpdate mixin.
+    """
+    config1 = copy.deepcopy(config)
+    config2 = copy.deepcopy(config)
+    config2[CONF_PREFIX][PREFIX_CMND] = "cmnd2"
+    data1 = json.dumps(config1)
+    data2 = json.dumps(config2)
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data1)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(f"{domain}.test")
+    assert state is not None
+    assert state.name == "Test"
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data1)
+    await hass.async_block_till_done()
+
+    assert not discovery_update.called
+
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data2)
+    await hass.async_block_till_done()
+
+    assert discovery_update.called
+
+
+async def help_test_discovery_device_remove(hass, mqtt_mock, domain, config):
+    """Test domain entity is removed when device is removed."""
+    device_reg = await hass.helpers.device_registry.async_get_registry()
+    entity_reg = await hass.helpers.entity_registry.async_get_registry()
+
+    config = copy.deepcopy(config)
+    unique_id = f"{config[CONF_MAC]}_{domain}_0"
+
+    data = json.dumps(config)
+    async_fire_mqtt_message(hass, f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config", data)
+    await hass.async_block_till_done()
+
+    device = device_reg.async_get_device(set(), {("mac", config[CONF_MAC])})
+    assert device is not None
+    assert entity_reg.async_get_entity_id(domain, "tasmota", unique_id)
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)